verifiers/internal/gha/verifier.go

```diff
@@ -80,11 +73,76 @@ func (v *GHAVerifier) VerifyArtifact(ctx context.Context,
 	return r, builderID, err
 }
 
+// VerifyArtifact verifies provenance for an artifact.
+func (v *GHAVerifier) VerifyArtifact(ctx context.Context,
+	provenance []byte, artifactHash string,
+	provenanceOpts *options.ProvenanceOpts,
+	builderOpts *options.BuilderOpts,
+) ([]byte, string, error) {
+	rClient, err := rekor.NewClient(defaultRekorAddr)
+	if err != nil {
+		return nil, "", err
+	}
+
+	/* Verify signature on the intoto attestation. */
+	env, cert, err := VerifyProvenanceSignature(ctx, rClient, provenance, artifactHash)
+	if err != nil {
+		return nil, "", err
+	}
+
+	return verifyEnvAndCert(env, cert,
+		provenanceOpts, builderOpts,
+		defaultArtifactTrustedReusableWorkflows)
+}
+
 // VerifyImage verifies provenance for an OCI image.
 func (v *GHAVerifier) VerifyImage(ctx context.Context,
-	provenance []byte, artifactHash string,
+	artifactImage string,
 	provenanceOpts *options.ProvenanceOpts,
 	builderOpts *options.BuilderOpts,
 ) ([]byte, string, error) {
-	return nil, "todo", serrors.ErrorNotSupported
+	/* Retrieve any valid signed attestations that chain up to Fulcio root CA. */
+	roots, err := fulcio.GetRoots()
+	if err != nil {
+		return nil, "", err
+	}
+	opts := &cosign.CheckOpts{
+		RootCerts: roots,
+	}
+
+	atts, _, err := container.RunCosignImageVerification(ctx,
+		artifactImage, opts)
+	if err != nil {
+		return nil, "", err
+	}
+
+	/* Now verify properties of the attestations */
+	var verifyErr error
+	var builderID string
+	var verifiedProvenance []byte
+	for _, att := range atts {
+		pyld, err := att.Payload()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "unexpected error getting payload from OCI registry %s", err)
+			continue
+		}
+		env, err := EnvelopeFromBytes(pyld)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "unexpected error parsing envelope from OCI registry %s", err)
+			continue
+		}
+		cert, err := att.Cert()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "unexpected error getting certificate from OCI registry %s", err)
+			continue
+		}
+		verifiedProvenance, builderID, verifyErr = verifyEnvAndCert(env,
+			cert, provenanceOpts, builderOpts,
+			defaultContainerTrustedReusableWorkflows)
+		if verifyErr == nil {
+			return verifiedProvenance, builderID, nil
```

Comment:
does this mean we're always returning a maximum of 1 provenance? This means it's fine to print only a single provenance when passing `-print-provenance`.

Reply:
Yes, correct. Updating. Will also create an issue to support multiple attestations returned on the container.